openwisp_monitoring/db/backends/influxdb2/client.py

```diff
@@ -0,0 +1,462 @@
+        if fields != '*':
+            if extra_fields and extra_fields != '*':
+                all_fields = [fields] + extra_fields if isinstance(extra_fields, list) else [fields, extra_fields]
+                field_filter = ' or '.join([f'r["_field"] == "{field}"' for field in all_fields])
+            else:
+                field_filter = f'r["_field"] == "{fields}"'
+            flux_query += f'\n  |> filter(fn: (r) => {field_filter})'
+
+        # Filter by tags
+        if tags:
+            tag_filters = ' and '.join([f'r["{tag}"] == "{value}"' for tag, value in tags.items()])
+            flux_query += f'\n  |> filter(fn: (r) => {tag_filters})'
+
+        # Add ordering
+        if order:
+            if order in ['time', '-time']:
+                desc = 'true' if order == '-time' else 'false'
+                flux_query += f'\n  |> sort(columns: ["_time"], desc: {desc})'
+            else:
+                raise self.client_error(
+                    f'Invalid order "{order}" passed.\nYou may pass "time" / "-time" to get '
+                    'result sorted in ascending /descending order respectively.'
+                )
+
+        # Add limit
+        if limit:
+            flux_query += f'\n  |> limit(n: {limit})'
+
+        # Pivot the result to make it similar to InfluxDB 1.x output
+        flux_query += '\n  |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")'
+
+        # Execute the query
+        try:
+            result = self.query_api.query(flux_query)
+            return self._parse_read_result(result)
+        except Exception as e:
+            logger.error(f"Error executing read query: {e}")
+            return []
+
+    def _parse_read_result(self, result):
+        parsed_result = []
+        for table in result:
+            for record in table.records:
+                parsed_record = {
+                    'time': record.get_time().isoformat(),
+                }
+                for key, value in record.values.items():
+                    if key not in ['_time', '_start', '_stop', '_measurement']:
+                        parsed_record[key] = value
+                parsed_result.append(parsed_record)
+        return parsed_result
+
+    def get_ping_data_query(self, bucket, start, stop, device_ids):
+        device_filter = ' or '.join([f'r["object_id"] == "{id}"' for id in device_ids])
+        query = f'''
+        from(bucket: "{bucket}")
+          |> range(start: {start}, stop: {stop})
+          |> filter(fn: (r) => r["_measurement"] == "ping")
+          |> filter(fn: (r) => r["_field"] == "loss" or r["_field"] == "reachable" or r["_field"] == "rtt_avg" or r["_field"] == "rtt_max" or r["_field"] == "rtt_min")
+          |> filter(fn: (r) => r["content_type"] == "config.device")
+          |> filter(fn: (r) => {device_filter})
+          |> aggregateWindow(every: v.windowPeriod, fn: mean, createEmpty: false)
+          |> yield(name: "mean")
+        '''
+        return query
```

Comment:
Please move this query to queries.py and use `get_query()` method to generate the final query.